segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/BaseChunkStorageProvider.java

```diff
@@ -0,0 +1,531 @@
+
+        // Call concrete implementation.
+        ChunkInfo info = doGetInfo(chunkName);
+
+        LoggerHelpers.traceLeave(log, "getInfo", traceId, chunkName);
+
+        return info;
+    }
+
+    /**
+     * Reads a range of bytes from the underlying chunk.
+     *
+     * @param handle       ChunkHandle of the chunk to read from.
+     * @param fromOffset   Offset in the chunk from which to start reading.
+     * @param length       Number of bytes to read.
+     * @param buffer       Byte buffer to which data is copied.
+     * @param bufferOffset Offset in the buffer at which to start copying read data.
+     * @return int Number of bytes read.
+     * @throws ChunkStorageException Throws ChunkStorageException in case of I/O related exceptions.
+     * @throws IllegalArgumentException If argument is invalid.
+     * @throws IndexOutOfBoundsException If the index is out of bounds.
+     */
+    @Override
+    final public int read(ChunkHandle handle, long fromOffset, int length, byte[] buffer, int bufferOffset) throws ChunkStorageException, NullPointerException, IndexOutOfBoundsException {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        // Validate parameters
+        Preconditions.checkArgument(null != handle, "handle");
+        Preconditions.checkArgument(null != buffer, "buffer");
+        Preconditions.checkArgument(fromOffset >= 0, "fromOffset must be non-negative");
+        Preconditions.checkArgument(length >= 0 && length <= buffer.length, "length");
+        Preconditions.checkElementIndex(bufferOffset, buffer.length, "bufferOffset");
+
+        long traceId = LoggerHelpers.traceEnter(log, "read", handle.getChunkName(), fromOffset, bufferOffset, length);
+        Timer timer = new Timer();
+
+        // Call concrete implementation.
+        int bytesRead = doRead(handle, fromOffset, length, buffer, bufferOffset);
+
+        Duration elapsed = timer.getElapsed();
+        ChunkStorageProviderMetrics.READ_LATENCY.reportSuccessEvent(elapsed);
+        ChunkStorageProviderMetrics.READ_BYTES.add(bytesRead);
+
+        log.debug("Read - chunk={}, offset={}, bytesRead={}, latency={}.", handle.getChunkName(), fromOffset, length, elapsed.toMillis());
+        LoggerHelpers.traceLeave(log, "read", traceId, bytesRead);
+
+        return bytesRead;
+    }
+
+    /**
+     * Writes the given data to the underlying chunk.
+     *
+     * @param handle ChunkHandle of the chunk to write to.
+     * @param offset Offset in the chunk to start writing.
+     * @param length Number of bytes to write.
+     * @param data   An InputStream representing the data to write.
+     * @return int Number of bytes written.
+     * @throws ChunkStorageException Throws ChunkStorageException in case of I/O related exceptions.
+     */
+    @Override
+    final public int write(ChunkHandle handle, long offset, int length, InputStream data) throws ChunkStorageException {
```

Comment:
what happens in case that data returns less than _length_ bytes ?
what happens in case that offset points to a part of the chunk that has already be written ?
is it supposed to be overwritten ?
or is it only expected that we are appending data to the end of the chunk ?

Reply:
1. It can write less bytes than length. In this case ChunkStorageManeger will make another call.
2. The writes are supposed to create same effect as if they are append only. The upper layers will guarantee that either data is never overwritten or that same data is written back at same offset.